File: projects/05-metrics-pipeline/src/pipeline.rs
//! The consumer pipeline: consume → roll up → batch-sink → fan out. (Wiring.)
//!
//! This is glue — it *drives* the verticals and ties them into a lifecycle. It
//! pulls raw lines off the durable stream, parses them (V1), folds them into the
//! rollup engine (V2), flushes closed windows to the batched ClickHouse sink (V3)
//! and the live SSE feed (V4), and acks the broker **only after** a batch is
//! durably written (at-least-once). With the verticals still `todo!()`, the loop
//! panics on its first real point — that panic is the worklist. It runs only when
//! `RUN_CONSUMER=true` (see `main.rs`), so the bare scaffold serves ingest cleanly.

use std::time::Duration;

use async_nats::jetstream::{self, Context};
use futures_util::StreamExt;
use tokio::sync::watch;
use tracing::{debug, error, info, warn};

use crate::parse;
use crate::rollup::Rollup;
use crate::sink::Sink;
use crate::sse::LiveFeed;

/// Tuning for the consumer pipeline, assembled in `main.rs`.
pub struct PipelineConfig {
    /// JetStream stream name to bind the durable consumer to.
    pub stream_name: String,
    /// Durable consumer name — survives restarts so the offset is remembered.
    pub durable_name: String,
    /// How often to run the watermark flush + time-based batch flush.
    pub flush_interval: Duration,
}

/// Run the pipeline until shutdown is signalled.
///
/// Owns the `Rollup` engine and the `Sink` for its lifetime; `LiveFeed` is the
/// shared broadcast hub (also held by the SSE handlers).
pub async fn run(
    js: Context,
    cfg: PipelineConfig,
    mut rollup: Rollup,
    mut sink: Sink,
    feed: LiveFeed,
    mut shutdown: watch::Receiver<bool>,
) {
    info!(stream = %cfg.stream_name, durable = %cfg.durable_name, "consumer pipeline starting");

    // --- Bind a durable pull consumer to the raw-metrics stream (wiring). ---
    let mut messages = match setup_consumer(&js, &cfg).await {
        Ok(m) => m,
        Err(e) => {
            error!(error = %e, "failed to create JetStream consumer; pipeline aborting");
            return;
        }
    };

    let mut flush_ticker = tokio::time::interval(cfg.flush_interval);

    loop {
        tokio::select! {
            // A new message from the durable stream.
            next = messages.next() => {
                let Some(msg) = next else {
                    warn!("consumer stream ended");
                    break;
                };
                match msg {
                    Ok(msg) => process_message(&mut rollup, &mut sink, &msg).await,
                    Err(e) => error!(error = %e, "consumer recv error"),
                }
            }

            // Periodic: close windows whose watermark passed, push to sink + feed,
            // and honour the sink's time-based flush trigger.
            _ = flush_ticker.tick() => {
                flush_windows(&mut rollup, &mut sink, &feed).await;
                if let Err(e) = sink.flush().await {
                    error!(error = %e, "time-triggered sink flush failed");
                }
            }

            // Graceful shutdown: drain partial windows, do a final flush, exit.
            _ = shutdown.changed() => {
                info!("pipeline draining on shutdown");
                flush_windows(&mut rollup, &mut sink, &feed).await; // TODO(V2): drain_all here
                if let Err(e) = sink.flush().await {
                    error!(error = %e, "final sink flush failed");
                }
                break;
            }
        }
    }

    info!("consumer pipeline stopped");
}

/// Decode one message into points (V1), fold them into the rollup engine (V2),
/// and ack the broker. NOTE the ack ordering is the at-least-once contract (V3):
/// a real implementation must ack only once the resulting rollups are durably
/// written — see SPEC V3.
async fn process_message(rollup: &mut Rollup, sink: &mut Sink, msg: &jetstream::Message) {
    let _ = sink; // the durable write + ack ordering is the V3 lesson (see SPEC).

    let body = String::from_utf8_lossy(&msg.payload);
    // V1: re-parse the durable line into points. (You could instead publish
    // already-parsed points to skip this; that's a design choice — SPEC V1.)
    match parse::parse(&body) {
        Ok(points) => {
            for p in &points {
                rollup.ingest(p); // V2
            }
        }
        Err(e) => warn!(error = %e, "dropping unparseable message from stream"),
    }

    // TODO(V3): ack only AFTER the rollups these points feed have been flushed
    // and durably written — acking here (before the sink write) would turn a
    // crash into silent data loss. Move the ack into the flush path.
    if let Err(e) = msg.ack().await {
        error!(error = %e, "failed to ack message");
    }
}

/// Flush closed windows (V2) into the batched sink (V3) and the live feed (V4).
async fn flush_windows(rollup: &mut Rollup, sink: &mut Sink, feed: &LiveFeed) {
    let rows = rollup.flush_ready(chrono::Utc::now()); // V2
    if rows.is_empty() {
        return;
    }
    debug!(count = rows.len(), "flushing closed windows");

    for row in &rows {
        feed.publish(row.clone()); // V4: live fan-out (drop-tolerant)
    }
    if let Err(e) = sink.push(rows).await {
        // V3: must NOT drop on the durable path — log and let redelivery retry.
        error!(error = %e, "sink push failed; rollups will be retried via redelivery");
    }
}

/// Create the durable pull consumer and return its message stream. (Wiring.)
async fn setup_consumer(
    js: &Context,
    cfg: &PipelineConfig,
) -> anyhow::Result<jetstream::consumer::pull::Stream> {
    let stream = js.get_stream(&cfg.stream_name).await?;
    let consumer = stream
        .get_or_create_consumer(
            &cfg.durable_name,
            jetstream::consumer::pull::Config {
                durable_name: Some(cfg.durable_name.clone()),
                // Explicit ack is what makes at-least-once possible (V3): an
                // unacked message is redelivered after its ack-wait elapses.
                ack_policy: jetstream::consumer::AckPolicy::Explicit,
                ..Default::default()
            },
        )
        .await?;
    Ok(consumer.messages().await?)
}

File: projects/05-metrics-pipeline/src/metrics_pipeline/pipeline.py
"""The consumer pipeline: consume -> roll up -> batch-sink -> fan out. (Wiring.)

This is glue — it *drives* the verticals and ties them into a lifecycle. It pulls
raw lines off the durable stream, parses them (V1), folds them into the rollup
engine (V2), flushes closed windows to the batched ClickHouse sink (V3) and the
live SSE feed (V4), and acks the broker **only after** a batch is durably written
(at-least-once). With the verticals still raising, the loop dies on its first
real point — that traceback is the worklist. It runs only when
`RUN_CONSUMER=true` (see `main.py`), so the bare scaffold serves ingest cleanly.

**Why one task and not several.** The loop is a single coroutine that owns the
`Rollup` and the `Sink` outright. That is what lets both be plain synchronous
objects with no locks: there is exactly one writer. The alternative — a fan-out
of consumer tasks sharing a rollup map — buys you nothing on CPython (the GIL
means the aggregation itself never runs in parallel) and costs you every
read-modify-write race in the engine. If you later need more throughput, the
scaling axis is *processes* (one consumer per shard of the series space), not
tasks sharing state. Write that decision down; it is a design-doc item.

The fetch timeout doubles as the loop's clock: a quiet stream wakes up every
`flush_interval` anyway, which is exactly when the watermark and the time-based
batch trigger need to run.
"""

from __future__ import annotations

import asyncio
import contextlib
from dataclasses import dataclass
from datetime import UTC, datetime

import structlog
from nats.aio.msg import Msg
from nats.errors import TimeoutError as NatsTimeoutError
from nats.js import JetStreamContext

from . import parse
from .rollup import Rollup
from .sink import Sink
from .sse import LiveFeed

__all__ = ["PipelineConfig", "run"]

log = structlog.get_logger(__name__)


@dataclass(slots=True)
class PipelineConfig:
    """Tuning for the consumer pipeline, assembled in `main.py`."""

    stream_name: str
    """JetStream stream to bind the durable consumer to."""
    subject: str
    durable_name: str
    """Durable consumer name — survives restarts so the offset is remembered."""
    fetch_batch: int
    """Messages pulled per fetch — the broker -> rollup prefetch bound."""
    flush_interval: float
    """Seconds between watermark flushes / time-based batch flushes."""


async def run(
    js: JetStreamContext,
    cfg: PipelineConfig,
    rollup: Rollup,
    sink: Sink,
    feed: LiveFeed,
    shutdown: asyncio.Event,
) -> None:
    """Run the pipeline until `shutdown` is set.

    Owns the `Rollup` and the `Sink` for its lifetime; `LiveFeed` is the shared
    hub, also held by the SSE handlers.
    """
    log.info("consumer pipeline starting", stream=cfg.stream_name, durable=cfg.durable_name)

    subscription = await js.pull_subscribe(
        cfg.subject,
        durable=cfg.durable_name,
        stream=cfg.stream_name,
    )
    try:
        await _consume(subscription, cfg, rollup, sink, feed, shutdown)
    finally:
        # Hand the subscription back, whatever ended the loop — a clean stop, a
        # cancellation, or a vertical's NotImplementedError. Skipping this is not
        # cosmetic: the connection's `drain()` in `main.py` waits on every live
        # subscription, so an orphaned one turns every shutdown into a 30-second
        # hang and then a SIGKILL from the orchestrator. The *durable consumer*
        # itself stays on the server (that is what remembers the offset); only
        # this client's interest in it goes away.
        with contextlib.suppress(Exception):
            await subscription.unsubscribe()
        log.info("consumer pipeline stopped")


async def _consume(
    subscription: JetStreamContext.PullSubscription,
    cfg: PipelineConfig,
    rollup: Rollup,
    sink: Sink,
    feed: LiveFeed,
    shutdown: asyncio.Event,
) -> None:
    """The loop itself, split out so `run` can own the subscription's lifetime."""
    while not shutdown.is_set():
        try:
            messages = await subscription.fetch(cfg.fetch_batch, timeout=cfg.flush_interval)
        except NatsTimeoutError:
            # No traffic this tick. Not an error — fall through to the flush so
            # an idle pipeline still closes its windows on schedule.
            messages = []
        except Exception as exc:  # noqa: BLE001 - a broker blip must not kill the loop
            log.error("consumer fetch failed", error=str(exc))
            await asyncio.sleep(cfg.flush_interval)
            continue

        for msg in messages:
            await process_message(rollup, msg)

        await flush_windows(rollup, sink, feed)
        try:
            await sink.flush()
        except Exception as exc:  # noqa: BLE001 - retried via redelivery (V3)
            log.error("time-triggered sink flush failed", error=str(exc))

    # Graceful shutdown: drain partial windows, do a final flush, exit.
    log.info("pipeline draining on shutdown")
    # TODO(V2 / graceful shutdown): use `rollup.drain_all()` here instead, so a
    # clean stop flushes the windows whose watermark hasn't passed yet rather
    # than dropping them. A crash may lose them (at-least-once covers it); a
    # deliberate stop should not.
    await flush_windows(rollup, sink, feed)
    try:
        await sink.flush()
    except Exception as exc:  # noqa: BLE001
        log.error("final sink flush failed", error=str(exc))


async def process_message(rollup: Rollup, msg: Msg) -> None:
    """Decode one message into points (V1) and fold them into the engine (V2).

    NOTE the ack ordering: acking here — before the rollups these points feed
    have been written — turns a crash into silent data loss. Getting it right is
    the V3 lesson.
    """
    try:
        points = parse.parse(msg.data.decode("utf-8", errors="strict"))
    except NotImplementedError:
        raise
    except Exception as exc:  # noqa: BLE001 - a poison message must not stop the stream
        log.warning("dropping unparseable message from stream", error=str(exc))
        await msg.ack()
        return

    for point in points:
        rollup.ingest(point)

    # TODO(V3): ack only AFTER the rollups these points feed have been flushed
    # and durably written. Move the ack into the flush path: hold the messages
    # whose points are still in open windows, and ack them once the batch
    # containing those windows lands in ClickHouse. Until then, redelivery is
    # your only safety net and this ack cancels it.
    await msg.ack()


async def flush_windows(rollup: Rollup, sink: Sink, feed: LiveFeed) -> None:
    """Flush closed windows (V2) into the batched sink (V3) and the feed (V4)."""
    rows = rollup.flush_ready(datetime.now(UTC))
    if not rows:
        return
    log.debug("flushing closed windows", count=len(rows))

    for row in rows:
        feed.publish(row)  # V4: live fan-out, drop-tolerant

    try:
        await sink.push(rows)
    except Exception as exc:  # noqa: BLE001
        # V3: must NOT drop on the durable path — log and let redelivery retry.
        log.error("sink push failed; rollups will be retried via redelivery", error=str(exc))
